alpa/create_state_parallel.py

```diff
@@ -0,0 +1,48 @@
+"""Compile executables for creating training state distributedly."""
+from jax._src.lib import xla_bridge as xb, xla_extension as xe
+from jax.core import (Jaxpr, ClosedJaxpr, Literal, new_jaxpr_eqn, gensym,
+                      get_aval, raise_to_shaped, AbstractValue)
+from jax.interpreters import partial_eval as pe
+from jax.tree_util import tree_flatten, tree_unflatten, PyTreeDef
+
+from alpa.shard_parallel.auto_sharding import run_spmd_partitioner_pass
+from alpa.mesh_executable import NormalMeshDriverExecutable
+from alpa.measure_record import StrategyConfig
+from alpa.util import jaxpr_to_hlo_module
+
+
+def compile_create_state_executable(fun, in_tree, out_tree_thunk, static_argnums,
+                                    donated_invars, batch_invars, train_step,
+                                    other_args, *avals):
+    # Trace to get jaxpr
+    jaxpr, out_avals, consts = pe.trace_to_jaxpr_final(fun, avals)
+    out_tree = out_tree_thunk()
+    state_aval = tree_unflatten(out_tree, out_avals)
+
+    executable = train_step.get_executable(state_aval, other_args)
+    placement_specs = executable.get_placement_specs()[0]
```

Comment:
Here do we assume that the output of `fun` should have exactly the same shape as `train_step`'s 0-th input argument? If so, we can add a document that explicitly says that.

Reply:
Will do